python/singa/autograd.py

```diff
@@ -1286,25 +1287,26 @@ def set_params(self, **parameters):
 
 
 class _BatchNorm2d(Operation):
-    def __init__(self, handle, name=None):
+    def __init__(self, handle, running_mean, running_var, name=None):
         super(_BatchNorm2d, self).__init__(name)
         self.handle = handle
+        self.running_mean = running_mean.data
+        self.running_var = running_var.data
 
-    def forward(self, x, scale, bias, running_mean, running_var):
-        self.running_mean = running_mean
-        self.running_var = running_var
+    def forward(self, x, scale, bias):
         if training:
 
             if isinstance(self.handle, singa.CudnnBatchNormHandle):
                 y, mean, var = singa.GpuBatchNormForwardTraining(
-                    self.handle, x, scale, bias, running_mean, running_var
+                    self.handle, x, scale, bias, self.running_mean, self.running_var
```

Comment:
pls do the cpu check ```if isinstance(self.handle, singa.CpuBatchNormHandle):``` because if there is no GPU, then there will be error like Cudnn... is unknown.

Reply:
for all cudnn checks in this file, need to replace them with the Cpu check.

Reply:
I see. I will modify the codes.

Reply:
Now Both CPU and GPU can run, but I used "type" instead of "isinstance" because CudnnBatchNormHandle is a subclass of BatchNormHandle, so CudnnBatchNormHandle is considered as an instance of BatchNormHandle in isinstance(). For more explaination, See http://www.runoob.com/python/python-func-isinstance.html

Moreover, I have degubbed the cpu batchnorm in the following two aspects:
(i) the forward function of cpu batchnorm needed the initialization of running mean and var, otherwise when it access the block it returns error as accessing an non-initialized block. I fixed this by initializating the mean by 0 and the var by 1.
(ii) the backward function CpuBatchNormBackward does not exist, but there is another function called CpuBatchNormBackwardx (in the directory src/model/operation/batchnorm.cc). So I use this function by providing all the necessary arguments.

The program can run now, but I am doing a brief cifar10 training test on AWS (using c5.x4large with 16 cpu cores).

Reply:
added the cpu test of conv2d and batchnorm2d in test_operation.py and passed, but still waiting for the cifar10 training accuracy test

Reply:
The following is the resnet18 training using CPU on CIFAR10 in the first few epochs. 
```
Start intialization............
Epoch=0: 100%|████████████████████████████████████████████████████████████████████████| 1562/1562 [2:09:57<00:00,  5.03s/it]
Training loss = 2233.394769, training accuracy = 0.490297
Test accuracy = 0.636218
Epoch=1: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [2:10:00<00:00,  4.98s/it]
Training loss = 1474.432049, training accuracy = 0.666633
Test accuracy = 0.678986
Epoch=2: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [2:10:11<00:00,  5.00s/it]
Training loss = 1163.035850, training accuracy = 0.741717
Test accuracy = 0.738181
Epoch=3: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [2:10:31<00:00,  5.03s/it]
Training loss = 979.977119, training accuracy = 0.782570
Test accuracy = 0.800581
Epoch=4: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [2:10:10<00:00,  4.98s/it]
Training loss = 872.811802, training accuracy = 0.806098
Test accuracy = 0.813902
Epoch=5: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 1562/1562 [2:10:05<00:00,  4.99s/it]
Training loss = 782.525783, training accuracy = 0.826144
Test accuracy = 0.832232
```
The training loss decreases normally. Therefore seems the batch norm is working.